odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+        different spaces:
+
+        >>> diff_space = odl.rn((2, 3, 4))
+        >>> diff_space == space
+        False
+        >>> diff_space = odl.rn((2, 3), exponent=1)
+        >>> diff_space == space
+        False
+        >>> diff_space = odl.rn((2, 3), dtype='float32')
+        >>> diff_space == space
+        False
+        >>> space == object
+        False
+
+        A `NumpyTensorSpace` with the same properties is considered
+        equal:
+
+        >>> same_space = odl.NumpyTensorSpace((2, 3), dtype='float64')
+        >>> same_space == space
+        True
+        """
+        if other is self:
+            return True
+
+        return (TensorSpace.__eq__(self, other) and
+                self.weighting == other.weighting)
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), self.shape, self.dtype, self.order,
+                     self.weighting))
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``."""
+        if isinstance(indices, list):
+            newshape = [self.shape[i] for i in indices]
+        else:
+            newshape = self.shape[indices]
+        if (newshape != self.shape and
+                isinstance(self.weighting, ArrayWeighting)):
+            # Cannot propagate the weighting array since the shape
+            # will be wrong. There is no unique way to get a weighting
+            # array for the new space, so we take no weighting.
+            weighting = NumpyTensorSpaceNoWeighting(
+                self.exponent, self.weighting.dist_using_inner)
+        else:
+            weighting = self.weighting
+        return type(self)(newshape, self.dtype, self.order,
+                          weighting=weighting)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        if self.ndim == 1:
+            posargs = [self.size]
+        else:
+            posargs = [self.shape]
+
+        if self.is_real_space:
+            constructor_name = 'rn'
+        elif self.is_complex_space:
+            constructor_name = 'cn'
+        else:
+            constructor_name = 'tensor_space'
+
+        if (constructor_name == 'tensor_space' or
+                not self.is_numeric or
+                self.dtype != self.default_dtype(self.field)):
+            posargs.append(dtype_str(self.dtype))
+
+        optargs = [('order', self.order, 'A')]
+        inner_str = signature_string(posargs, optargs)
+        weight_str = self.weighting.repr_part
+        if weight_str:
+            inner_str += ', ' + weight_str
+
+        return '{}({})'.format(constructor_name, inner_str)
+
+    @property
+    def element_type(self):
+        """Type of elements in this space: `NumpyTensor`."""
+        return NumpyTensor
+
+
+class NumpyTensor(Tensor):
+
+    """Representation of a `NumpyTensorSpace` element."""
+
+    def __init__(self, space, data):
+        """Initialize a new instance."""
+        assert isinstance(space, NumpyTensorSpace)
```

Comment:
Again, I think  we should remove this asserts for performance reasons since they should only capture developer errors.